Detector/DetComponents/src/CreateCellPositions.cpp

```diff
@@ -0,0 +1,113 @@
+#include "DD4hep/Readout.h"
+#include "DD4hep/Volumes.h"
+#include "TGeoManager.h"
+
+// EDM
+#include "datamodel/CaloHitCollection.h"
+#include "datamodel/PositionedCaloHitCollection.h"
+#include "datamodel/PositionedTrackHitCollection.h"
+#include "datamodel/TrackHitCollection.h"
+
+typedef CreateCellPositions<fcc::CaloHitCollection, fcc::PositionedCaloHitCollection> CreateCellCaloPositions;
+typedef CreateCellPositions<fcc::TrackHitCollection, fcc::PositionedTrackHitCollection> CreateCellTrackPositions;
+DECLARE_ALGORITHM_FACTORY(CreateCellCaloPositions)
+DECLARE_COMPONENT_WITH_ID(CreateCellCaloPositions, "CreateCellCaloPositions")
+DECLARE_ALGORITHM_FACTORY(CreateCellTrackPositions)
+DECLARE_COMPONENT_WITH_ID(CreateCellTrackPositions, "CreateCellTrackPositions")
+
+template <class Hits, class PositionedHit>
+CreateCellPositions<Hits, PositionedHit>::CreateCellPositions(const std::string& name, ISvcLocator* svcLoc)
+    : GaudiAlgorithm(name, svcLoc), m_geoSvc("GeoSvc", "CreateCellPositions") {
+  declareProperty("hits", m_hits, "Hit collection (input)");
+  declareProperty("positionedHits", m_positionedHits, "Positions of centres of volumes (output)");
+  declareProperty("readoutName", m_readoutName);
+}
+
+template <class Hits, class PositionedHit>
+StatusCode CreateCellPositions<Hits, PositionedHit>::initialize() {
+  return GaudiAlgorithm::initialize();
+}
+
+template <class Hits, class PositionedHit>
+StatusCode CreateCellPositions<Hits, PositionedHit>::execute() {
+
+  // Get the input hit collection
+  const Hits* hits = m_hits.get();
+  debug() << "Input hit collection size: " << hits->size() << endmsg;
+
+  // Initialize output collection
+  auto edmPositionedHitCollection = m_positionedHits.createAndPut();
+
+  uint64_t cellid = 0;
+  DD4hep::Geometry::VolumeManager volman = m_geoSvc->lcdd()->volumeManager();
+  auto readout = m_geoSvc->lcdd()->readout(m_readoutName);
+
+  // Loop though hits, retrieve volume position from cellID
+  for (const auto& cell : *hits) {
+    cellid = cell.core().cellId;
+    auto detelement = volman.lookupDetElement(cellid);
+    const auto& transformMatrix = detelement.worldTransformation();
+    debug() << "VolID: " << detelement.volumeID() << endmsg;
+    debug() << "DetElement ID: " << detelement.id() << endmsg;
+    double outGlobal[3];
+    double inLocal[] = {0, 0, 0};
+    transformMatrix.LocalToMaster(inLocal, outGlobal);
+
+    debug() << "Position of volume (mm) : \t" << outGlobal[0] / dd4hep::mm << "\t" << outGlobal[1] / dd4hep::mm << "\t"
+            << outGlobal[2] / dd4hep::mm << endmsg;
+
+    // get PhiEta segmentation
+    DD4hep::DDSegmentation::GridPhiEta* segmentation;
+    segmentation = dynamic_cast<DD4hep::DDSegmentation::GridPhiEta*>(
+        m_geoSvc->lcdd()->readout(m_readoutName).segmentation().segmentation());
+    if (segmentation == nullptr) {
+      error() << "There is no phi-eta segmentation!!!!" << endmsg;
+      return StatusCode::FAILURE;
+    }
```

Comment:
This part (retrieving the segmentation) could be moved in the initialization, not to be called for each event & each hit.